### PR TITLE
IX Bid Adapter: Multiformat support, video adunit param, IXdiags

### PR DESCRIPTION
**Pull request by umakajan**
## Type of change

-  Bugfix
-  Feature
-  New bidder adapter
-  Code style update (formatting, local variables)
- [ x ] Refactoring (no functional changes, no api changes)
-  Build related changes
-  CI related changes
-  Does this change affect user-facing APIs or examples documented on [http://prebid.org](http://prebid.org/)?
-  Other

## Description of change

- Added Multi format Ad unit support
- Update to diagnostics
- Reading video params from Adunit level

For any changes that affect user-facing APIs or example code documented on [http://prebid.org](http://prebid.org/), please provide:

- A link to a PR on the docs repo at [prebid/prebid.github.io#2494](https://github.com/prebid/prebid.github.io/pull/2494)

### CODE REVIEW

File: modules/ixBidAdapter.js [hunk 1]
```diff
@@ -285,6 +299,14 @@ function buildRequest(validBidRequests, bidderRequest, impressions, version) {
   r.ext.source = 'prebid';
   r.ext.ixdiag = {};
 
+  // getting ixdiags for adunits of the video, outstream & multi format (MF) style
+  let ixdiagMF = buildMultiFormatDiag(validBidRequests);
+  for (var key in ixdiagMF) {
+    if (ixdiagMF.hasOwnProperty(key)) {
```
Comment (msm0504) [inline modules/ixBidAdapter.js:305]: Why is this check needed? What properties will be in `ixdiagMF` with `hasOwnProperty` returning `false`?

File: modules/ixBidAdapter.js [hunk 2]
```diff
@@ -560,32 +673,59 @@ export const spec = {
...
       return false;
     }
 
-    if (bid.hasOwnProperty('mediaType') && !(utils.contains(SUPPORTED_AD_TYPES, bid.mediaType))) {
+    if (bid.hasOwnProperty('mediaTypes') && !(mediaTypeBannerSizes || mediaTypeVideoPlayerSize)) {
       return false;
     }
 
-    if (bid.hasOwnProperty('mediaTypes') && !(utils.deepAccess(bid, 'mediaTypes.banner.sizes') || utils.deepAccess(bid, 'mediaTypes.video.playerSize'))) {
+    if (!includesSize(bid.sizes, bid.params.size) &&
```
Comment (msm0504) [inline modules/ixBidAdapter.js:697]: Use either `bid.params.size` or `paramsSize `. Don't mix both in the same `if ` condition

File: modules/ixBidAdapter.js [hunk 3]
```diff
@@ -466,6 +488,96 @@ function buildRequest(validBidRequests, bidderRequest, impressions, version) {
...
+ * @param {array} validBidRequests  The bid requests from prebid
+ * @return {Object} IX diag values for ad units
+ */
+function buildMultiFormatDiag(validBidRequests) {
+  var adUnitMap = {};
+  // create ad unit map and collect the required diag properties
+  for (let i = 0; i < validBidRequests.length; ++i) {
+    var bid = validBidRequests[i];
+    var trId = bid.transactionId;
+    if (!adUnitMap.hasOwnProperty(trId)) {
```
Comment (msm0504) [inline modules/ixBidAdapter.js:504]: Can multiple ad units have the same transaction id? If not, remove this check.

File: modules/ixBidAdapter.js [hunk 4]
```diff
@@ -466,6 +488,96 @@ function buildRequest(validBidRequests, bidderRequest, impressions, version) {
...
+    bu: 0,
+    iu: 0,
+    nu: 0,
+    ou: 0,
+    allU: 0
+  };
+
+  // Iterate over adunit map and populate MF diag object
+  const adIds = Object.keys(adUnitMap);
+  for (let i = 0; i < adIds.length; ++i) {
```
Comment (msm0504) [inline modules/ixBidAdapter.js:550]: If each ad unit has a unique transaction id, increasing these counts can be included in the loop that populates `adUnitMap` instead of looping over the ad units again.

**Review by msm0504 — CHANGES_REQUESTED** [review]

File: modules/ixBidAdapter.js [hunk 5]
```diff
@@ -466,6 +486,66 @@ function buildRequest(validBidRequests, bidderRequest, impressions, version) {
...
 }
+
+/**
+ * Calculates IX diagnostics values and packages them into an object
+ *
+ * @param {array} validBidRequests  The valid bid requests from prebid
+ * @return {Object} IX diag values for ad units
+ */
+function buildIXDiag(validBidRequests) {
+  var adUnitMap = validBidRequests
```
Comment (msm0504) [inline modules/ixBidAdapter.js:497]: Why is this needed?

**Review by msm0504 — CHANGES_REQUESTED** [review]

File: modules/ixBidAdapter.js (hunk 5 above)
Comment (umakajan) [reply modules/ixBidAdapter.js:497]: We want to capture metrics for each ad unit for every unique `transactionId` from the valid bid requests

At a high level, we will be passing these metrics in our bid request for further analytics

**Review by umakajan — COMMENTED** [review]
@msm0504 we addressed the previous set of code review comments. Are you able to identify how we can resolve the CircleCI failing test for browserstack job? We already merged master into this branch. Let us know if you have any other feedback.

**Review by msm0504 — APPROVED** [review]
